### PR TITLE
Feature | Add support server SPN option

### PR DESCRIPTION
**Pull request by DavoudEshtehari**
Skip the driver-generated SPN in case of providing `Server SPN` and `Failover Partner SPN` options in a connection string.

- [✔️] Supports netfx ([native sni](https://sqlclientdrivers.visualstudio.com/ADO.Net/_git/Microsoft.Data.SqlClient.sni/pullrequest/3331) from the internal feed is required).
- [✔️] Supports .net core on Windows ([native sni](https://sqlclientdrivers.visualstudio.com/ADO.Net/_git/Microsoft.Data.SqlClient.sni/pullrequest/3331) from the internal feed is required).
- [✔️] Supports .net core on Linux (or Windows by enabling the [Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows](https://github.com/dotnet/SqlClient/blob/main/BUILDGUIDE.md#using-managed-sni-on-windows) switch).

link: [Service Principal Names](https://docs.microsoft.com/en-us/windows/win32/ad/service-principal-names)

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs [hunk 1]
```diff
@@ -138,14 +139,24 @@ private SNINativeMethodWrapper.ConsumerInfo CreateConsumerInfo(bool async)
...
         {
             // We assume that the loadSSPILibrary has been called already. now allocate proper length of buffer
             spnBuffer = new byte[1][];
             if (isIntegratedSecurity)
             {
                 // now allocate proper length of buffer
-                spnBuffer[0] = new byte[SNINativeMethodWrapper.SniMaxComposedSpnLength];
+                if (!string.IsNullOrEmpty(serverSPN))
+                {
+                    byte[] srvSPN = Encoding.Unicode.GetBytes(serverSPN);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectNative.cs:151]: How come in SNIProxy.cs, we are using UTF8 (`Encoding.UTF8.GetBytes(serverSPN)`) and here we are using Unicode.GetBytes?

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/StringsHelper.cs [hunk 2]
```diff
@@ -103,8 +103,9 @@ internal class ResourceNames
             internal const string DbConnectionString_DataSource = @"Indicates the name of the data source to connect to.";
             internal const string DbConnectionString_Encrypt = @"When true, SQL Server uses SSL encryption for all data sent between the client and server if the server has a certificate installed.";
             internal const string DbConnectionString_Enlist = @"Sessions in a Component Services (or MTS, if you are using Microsoft Windows NT) environment should automatically be enlisted in a global transaction where required.";
-            internal const string DbConnectionString_InitialCatalog = @"The name of the initial catalog or database in the data source.";
             internal const string DbConnectionString_FailoverPartner = @"The name or network address of the instance of SQL Server that acts as a failover partner.";
+            internal const string DbConnectionString_FailoverPartnerSPN = @"Indicate the service principal name (SPN) of the failover partner to connect to.";
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/StringsHelper.cs:107]:
```suggestion
            internal const string DbConnectionString_FailoverPartnerSPN = @"The service principal name (SPN) of the failover partner.";
```

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/StringsHelper.cs [hunk 3]
```diff
@@ -116,6 +117,7 @@ internal class ResourceNames
             internal const string DbConnectionString_PersistSecurityInfo = @"When false, security-sensitive information, such as the password, is not returned as part of the connection if the connection is open or has ever been in an open state.";
             internal const string DbConnectionString_Pooling = @"When true, the connection object is drawn from the appropriate pool, or if necessary, is created and added to the appropriate pool.";
             internal const string DbConnectionString_Replication = @"Used by SQL Server in Replication.";
+            internal const string DbConnectionString_ServerSPN = @"Indicate the service principal name (SPN) of the data source to connect to.";
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/StringsHelper.cs:120]:
```suggestion
            internal const string DbConnectionString_ServerSPN = @"The service principal name (SPN) of the server.";
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx [hunk 4]
```diff
@@ -4617,4 +4617,10 @@
   <data name="SQL_ParameterDirectionInvalidForOptimizedBinding" xml:space="preserve">
     <value>Parameter '{0}' cannot have Direction Output or InputOutput when EnableOptimizedParameterBinding is enabled on the parent command.</value>
   </data>
+  <data name="DbConnectionString_FailoverPartnerSPN" xml:space="preserve">
+    <value>Indicates the service principal name (SPN) of the failover partner to connect to.</value>
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx:4621]:
```suggestion
    <value>The service principal name (SPN) of the failover partner.</value>
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx [hunk 5]
```diff
@@ -4617,4 +4617,10 @@
...
+  </data>
+  <data name="DbConnectionString_ServerSPN" xml:space="preserve">
+    <value>Indicates the service principal name (SPN) of the data source to connect to.</value>
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.resx:4624]:
```suggestion
    <value>The service principal name (SPN) of the server.</value>
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs [hunk 6]
```diff
@@ -5694,6 +5694,15 @@ internal class Strings {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to Indicates the service principal name (SPN) of the failover partner to connect to..
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs:5698]:
```suggestion
        ///   Looks up a localized string similar to The service principal name (SPN) of the failover partner..
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs [hunk 7]
```diff
@@ -5847,6 +5856,15 @@ internal class Strings {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to Indicates the service principal name (SPN) of the data source to connect to..
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs:5860]:
```suggestion
        ///   Looks up a localized string similar to The service principal name (SPN) of the server..
```

**Review by David-Engel — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs [hunk 8]
```diff
@@ -1910,11 +1910,8 @@ private void ResolveExtendedServerName(ServerInfo serverInfo, bool aliasLookup,
                             this,
                             ignoreSniOpenTimeout,
                             timeout.LegacyTimerExpire,
-                            ConnectionOptions.Encrypt,
-                            ConnectionOptions.TrustServerCertificate,
-                            ConnectionOptions.IntegratedSecurity,
+                            ConnectionOptions,
                             withFailover,
-                            ConnectionOptions.Authentication,
                             ConnectionOptions.HostNameInCertificate);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:1915]: As long as we are collapsing these parameters into ConnectionOptions, we should include HostNameInCertificate, too.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs [hunk 9]
```diff
@@ -2298,13 +2298,9 @@ private void AttemptOneLogin(ServerInfo serverInfo, string newPassword, SecureSt
...
                             ignoreSniOpenTimeout,
                             timeout.LegacyTimerExpire,
-                            ConnectionOptions.Encrypt,
-                            ConnectionOptions.TrustServerCertificate,
-                            ConnectionOptions.IntegratedSecurity,
+                            ConnectionOptions,
                             withFailover,
                             isFirstTransparentAttempt,
-                            ConnectionOptions.Authentication,
-                            ConnectionOptions.Certificate,
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:2307]: Same comment here: As long as we are collapsing these parameters into ConnectionOptions, we should include HostNameInCertificate, too.

**Pull request merged**